### PR TITLE
XML Signing

### PR DESCRIPTION
**Pull request by Calpicow**
Only supports RSA keys. I can't seem to load ECDSA, and DSA is not really supported in Go. Piggybacks a lot from validator.go

### CODE REVIEW

**Comment by ma314smith** [comment]
Thanks for taking on the singing part of this.  I like where you going with it, but just merged in some changes to make things a little smoother.

- Rather than using the `Validator` directly, I've split out the functionality that you needed into `signatureData`, so that `Validator` and `Signer` can remain separate concerns.  Take a look at how `Validator` is using it, and let me know if that makes sense to you.

- Also, I changed some of the "set" methods to "parse" (ex. `setSigValue` became `parseSigValue`), since that seems more appropriate for validation, while setting would be done during the signing process

- Most of my experience is around the validation side, so I'm not sure of all the different signing flows, but it seems like maybe there should be 3 ways to sign:
1. The Signature element already exists in the xml, and the Digest/Sig Value just need to be set (this is what you already set up)
2. The Signature element is external, and gets set up separately
3. The Signature element gets created and inserted into the xml that's being signed

It seems like number 3 would be the most common/useful, but I could be wrong...I'm curious what your opinion is here.  Would it make sense to provide the xml that needs to be signed (without a Signature element already present), and then have a Sign method that takes parameters for the algorithms needed to create and insert the Signature?  Or create functions to `setSigAlgorithm`, etc, that can be used before calling Sign?

File: signer.go [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+	}
+
+	var hashed, signature []byte
+	var h1, h2 *big.Int
+	signingAlgorithm, ok := signingAlgorithms[s.validator.sigAlgorithm]
+	if !ok {
+		return errors.New("signedxml: unsupported algorithm")
+	}
+
+	switch signingAlgorithm.hash {
```
Comment (ma314smith) [inline signer.go]: No need to use the switch block here.  Just call New() on signingAlgorithm.hash:
```go
hasher := signingAlgorithm.hash.New()
hasher.Write([]byte(canonSignedInfo))
hashed = hasher.Sum(nil)
```

**Comment by Calpicow** [comment]
I think the biggest problem for 3 is that we don't know where to the XML to insert the signature. I've seen it both within the `Response` and `Assertion` elements when working with SAML (you're on your own for other standards). I've based the signing mechanism on how the `xmlsec1` command line tool works (method 1).

So if we want to implement 3, we could create a `SetSignatureParams` that would let you specify where to insert the signature, and the algorithms for canonicalization, signature, transform(s), and digest. This would require exposing `etree` in some form (or xpath), but I'm not sure if we want to do that.

2 would come pretty much for free once 3 is possible. We'll just expose a `GetSignature` method.

**Comment by ma314smith** [comment]
That makes sense. I'm fine with just method 1 for now if you can remove the validator field, and embed signatureData instead.  Also, would you mind adding a signing example to the ReadMe once you get it finalized?

**Comment by Calpicow** [comment]
Done

**Pull request merged**